part/node.go

```diff
@@ -0,0 +1,322 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package part
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"strings"
+	"unsafe"
+)
+
+type nodeKind uint8
+
+const (
+	nodeKindUnknown = iota
+	nodeKind4
+	nodeKind16
+	nodeKind48
+	nodeKind256
+)
+
+type leaf[T any] struct {
+	key   []byte
+	value T
+}
+
+type node[T any] struct {
+	flags  uint16 // kind(4b) | unused(3b) | size(9b)
+	prefix []byte
+	leaf   *leaf[T]
+	watch  chan struct{}
+}
+
+const kindMask = uint16(0b1111_000_00000000_0)
+
+func (n *node[T]) kind() nodeKind {
+	return nodeKind(n.flags >> 12)
+}
+
+func (n *node[T]) setKind(k nodeKind) {
+	n.flags = (n.flags & ^kindMask) | (uint16(k&0b1111) << 12)
+}
+
+const sizeMask = uint16(0b0000_000_1111_1111_1)
+
+func (n *node[T]) cap() int {
+	switch n.kind() {
+	case nodeKind4:
+		return 4
+	case nodeKind16:
+		return 16
+	case nodeKind48:
+		return 48
+	case nodeKind256:
+		return 256
+	default:
+		panic("unknown node kind")
+	}
+}
+
+func (n *node[T]) size() int {
+	return int(n.flags & sizeMask)
+}
+
+func (n *node[T]) setSize(size int) {
+	n.flags = (n.flags & ^sizeMask) | uint16(size)&sizeMask
+}
+
+func (n *node[T]) self() *node[T] {
+	return n
+}
+
+func (n *node[T]) node4() *node4[T] {
+	return (*node4[T])(unsafe.Pointer(n))
+}
+
+func (n *node[T]) node16() *node16[T] {
+	return (*node16[T])(unsafe.Pointer(n))
+}
+
+func (n *node[T]) node48() *node48[T] {
+	return (*node48[T])(unsafe.Pointer(n))
+}
+
+func (n *node[T]) node256() *node256[T] {
+	return (*node256[T])(unsafe.Pointer(n))
+}
+
+// clone returns a shallow clone of the node.
+// We are working on the assumption here that only
+// value-types are mutated in the returned clone.
+func (n *node[T]) clone(watch bool) *node[T] {
+	var nCopy *node[T]
+	switch n.kind() {
+	case nodeKind4:
+		n4 := *n.node4()
+		nCopy = (&n4).self()
+	case nodeKind16:
+		n16 := *n.node16()
+		nCopy = (&n16).self()
+	case nodeKind48:
+		n48 := *n.node48()
+		nCopy = (&n48).self()
+	case nodeKind256:
+		nCopy56 := *n.node256()
+		nCopy = (&nCopy56).self()
+	default:
+		panic("unknown node kind")
+	}
+	if watch {
+		nCopy.watch = make(chan struct{})
+	} else {
+		nCopy.watch = nil
+	}
+	if n.leaf != nil {
+		nCopy.leaf = &leaf[T]{
+			key:   n.leaf.key,
+			value: n.leaf.value,
+		}
+	}
+	return nCopy
+}
+
+func (n *node[T]) promote(watch bool) *node[T] {
```

Comment:
I went for simplicity so I didn't even initially specialize node256 (until I benchmarked and noticed it makes a ton of sense). Since the perf is already 2x to what we had before I would leave it at this for now.